### DISCUSSION TITLE
Suport for IPv6

### DISCUSSION THREAD

**Discussion by sorin-silaghi** [pivpn/pivpn#1394] [Feature requests]
Any chance we could get support for IPv6 in the config process? My ISP only provides IPv6 and I haven't been able to get the VPN to work yet.

**Comment by redfast00** [comment]
We don't have that yet, but that's a really good question. For now: do the install, and then afterwards edit the config files as described in https://community.openvpn.net/openvpn/wiki/IPv6 @0-kaladin: this is something we will need to do

**Comment by redfast00** [comment]
Change `proto udp` into `proto udp6`

**Comment by coolapso** [comment]
@nellsavedra if you think this is mandatory, just make a pull request yourself! 

sorry if I am being quite rude, but there's no such thing as mandatory and priority around here. 
this is mainly maintained with @0-kaladin and @redfast00 free time, and more recently by me, and I am also with a big lack of free time since I just changed job and country. 

so if you think this is URGENT, well do it and make a pull request, otherwise, you will have to wait until we have time for that! 

I am sorry.

**Comment by eiGelbGeek** [comment]
You only have to change one line in server.conf!
tcp to tcp6 or from upd to udp6!
In your client configurations, you just have to replace your IPv4 with your IPv6! 
Also change udp -> udp6 / tcp -> tcp6

If i find some time i create a pull request ... one little pull request is done ... that fixed the stats ...
A IPv6 is simply longer than a IPv4 :-)

**Comment by coolapso** [comment]
Considering:  https://github.com/pivpn/pivpn/issues/619
@orazioedoardo @devZer0

Shall we disable IPv6 until we support setup for it? 
actually, I don't think it will solve the issue as the connection probably falls back to IPv6 on the client side of things ... 
As i don't have IPv6 can't really test this ... any comments?

**Comment by mwoolweaver** [comment]
i actually just disabled ipv6 on my windows machine and everything works as it should. I know this isn't the preferred option but it works for me. 


My ISP is Comcast and the router hands out ipv6 by default but it _will_ do ipv4 as a fallback.

**Comment by coolapso** [comment]
That's the only doubt I have .. if disabling IPv6 on the server side will be enough to fix the issue ... but I expect that the client will eventually fall back to IPv6 if its enable on clientside thus bypassing the VPN and leaking the IPv6

**Comment by mwoolweaver** [comment]
I honestly think it client depended as my Asus router && Android phone do not have this issue and neither leak ipv6 DNS to my knowledge. 


I have tested with all sorts dns leak testing sites. And seen 0 positive results in that area.

**Comment by mwoolweaver** [comment]
It's possible to enable ipv6 networking over VPN, correct? If that's the case then it wouldn't fallback as long as blocking outside dns still happened? Im not completely sure the commands to setup and test a dual stack config are or I would test them

**Comment by MichaIng** [comment]
Not 100% sure how OpenVPN handles it, but in WireGuard, if one does not enable IPv6 forwarding on the server, there are two possible issues if the client supports IPv6:
- You force `::1` IPv6 route through the VPN on the client, then IPv6 connections simply hang.
- You do not force IPv6 through VPN, then those will bypass the VPN, so inexperienced users will never know if they have a private connection up or not, being an in-acceptable case.
- So either all clients need to disable IPv6 (and this is not always possible, e.g. in case or Android and other restrictive OS), or IPv6 forwarding/handling needs to be enabled on the server to have an acceptable situation.

As said not sure how this is/can be handled with OpenVPN. If there is a way to force IPv6 to IPv4 fallback silently without any hang, then this is probably the easiest solution for now.

Of course all this applies for DNS requests as well.

**Comment by ovz93br43v7** [comment]
> You only have to change one line in server.conf!
> tcp to tcp6 or from upd to udp6!
> In your client configurations, you just have to replace your IPv4 with your IPv6!
> Also change udp -> udp6 / tcp -> tcp6
> 
> If i find some time i create a pull request ... one little pull request is done ... that fixed the stats ...
> A IPv6 is simply longer than a IPv4 :-)

Is this explanation valid for openvpn or wireguard? If it is for openvpn, can you describe what I have to do in wireguard to enable the IPv6 support? Reason: I read that some hotels provide only IPv6 WLAN, then I assume the current configuration will not work?

**Comment by MichaIng** [comment]
> tcp to tcp6 or from upd to udp6!

This is for OpenVPN, although don't ask me about the details 😉.

#### For WireGuard:
- Assure that the server can connect to IPv6 hosts, e.g.: `ping -6 dns9.quad9.net`
- **Server-side**, enable forwarding for IPv6 separately and assure that the server accepts IPv6 router advertisements even that it acts as router for the VPN itself (`accept_ra=2`):
    ```
    PostUp = sysctl net.ipv6.conf.eth0.accept_ra=2
    PostUp = sysctl net.ipv6.conf.%i.forwarding=1 net.ipv6.conf.eth0.forwarding=1
   PostUp = ip6tables -A FORWARD -i %i -j ACCEPT; ip6tables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
   PostDown = ip6tables -D FORWARD -i %i -j ACCEPT; ip6tables -t nat -D POSTROUTING -o eth0 -j MASQUERADE
    ```
    - Forwarding rules be applied via `nftables` (or `bpfilter`?) as well, depending on machine and kernel capabilities + available userspace firewall frontends.
    - Replace `eth0` with the actual network interface for outgoing requests on the server.
- **Client-side** force IPv6 requests to be done through the VPN as well separately:
   ```
   AllowedIPs = 0.0.0.0/0, ::/0
   ```

**Comment by ovz93br43v7** [comment]
Thanks for your fast answer!
Just to be sure to understand everything:
>     * Assure that the server can connect to IPv6 hosts, e.g.: `ping -6 dns9.quad9.net`
Ok I have tired this on my raspberry pi, result:
`PING dns9.quad9.net(2620:fe::9) 56 data bytes
64 bytes from 2620:fe::9: icmp_seq=1 ttl=58 time=14.2 ms`
I would assume this is ok because "2620:fe::9" looks like a IPv6 address.

This 4 lines has to be added to iptables (this is installed on raspbian) and modified according to my eth0 interface?:
>     * **Server-side**, enable forwarding for IPv6 separately and assure that the server accepts IPv6 router advertisements even that it acts as router for the VPN itself (`accept_ra=2`):
>  .....

As far as I understand this was already added by pivpn during setup, I can see this in my client configuration details:
>     * **Client-side** force IPv6 requests to be done through the VPN as well separately:
>       ```
>       AllowedIPs = 0.0.0.0/0, ::/0
>       ```

**Comment by MichaIng** [comment]
> This 4 lines has to be added to iptables (this is installed on raspbian) and modified according to my eth0 interface?

The 4 lines have to be added to your WireGuard server config `/etc/wireguard/wg0.conf` to the `[Interface]` block. So on startup via wg-quick service, iptables is called automatically to establish the forwarding rules.

> As far as I understand this was already added by pivpn during setup, I can see this in my client configuration details:

Ah yes, I should have checked what PiVPN actually adds 😄.

**Comment by ovz93br43v7** [comment]
@MichaIng Thanks again for your fast answer, fantastic support!
I have added the four lines to the wg0.conf file and tested with ipv6-test.com whether IPv6 support is detected using wireguard VPN on my smartphone. But it still shows IPv6 not supported.

>       * Replace `eth0` with the actual network interface for outgoing requests on the server.

My RPi is connected with a ethernet cable and is using just ethernet connection, so I left `eth0` in the four lines. Or is the "interface for outgoing requsts on the server" the wiregurard (wg0) interface?!?

**Comment by orazioedoardo** [comment]
> I have added the four lines to the wg0.conf file and tested with ipv6-test.com whether IPv6 support is detected using wireguard VPN on my smartphone. But it still shows IPv6 not supported.

Do you have IPv6 connectivity on your cellular connection?

**Comment by ovz93br43v7** [comment]
You are right I don't have IPv6 connectivity on my cellular connection. Sorry I missed that...

**Comment by ovz93br43v7** [comment]
> The 4 lines have to be added to your WireGuard server config `/etc/wireguard/wg0.conf` to the `[Interface]` block. So on startup via wg-quick service, iptables is called automatically to establish the forwarding rules.

Is it normal that after adding these four lines ifconfig shows me for eth0:
two `inet6 fd00`, two `inet6 200x` addresses and one `fe80` address? (before adding just one `fd00`, `fe80`, `200x` address)?

**Comment by MichaIng** [comment]
Which network interface(s) are those addresses attached to? `ip a`

**Comment by ovz93br43v7** [comment]
> Which network interface(s) are those addresses attached to? `ip a`

As far as I understand to eth0, here the output of `ip a`


```
XXXXXX@raspberrypi:~ $ ip a

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000

	link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
	inet xxx.xxx.xxx.xxx/8 scope host lo 
		valid_lft forever preferred_lft forever
	inet6 ::1/128 scope host 
	       valid_lft forever preferred_lft forever

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000

	link/ether xx:xx:xx:xx:xx:xx brd xx:xx:xx:xx:xx:xx
	inet xxx.xxx.xxx.xxx/xx brd xxx.xxx.xxx.xxx scope global noprefixroute eth0
		valid_lft forever preferred_lft forever
	inet6 200x:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx/64 scope global dynamic mngtmpaddr 	noprefixroute 
		valid_lft 7154sec preferred_lft 1216sec
	inet6 200x:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx/64 scope global dynamic mngtmpaddr 
		valid_lft 7154sec preferred_lft 1216sec
	inet6 fd00::xxxx:xxxx:xxxx:xxxx/64 scope global dynamic mngtmpaddr 
		valid_lft 7154sec preferred_lft 3554sec
	inet6 fd00::xxxx:xxxx:xxxx:xxxx/64 scope global dynamic mngtmpaddr noprefixroute 
		valid_lft 7154sec preferred_lft 3554sec
	inet6 fe80::xxxx:xxxx:xxxx:xxxx/64 scope link 
		valid_lft forever preferred_lft forever

3: wlan0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
	link/ether xx:xx:xx:xx:xx:xx brd ff:ff:ff:ff:ff:ff

4: wg0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
	link/none 
	inet xxx.xxx.xxx.xxx/24 scope global wg0
		valid_lft forever preferred_lft forever
```

EDIT: I tried today in a WiFi of a friend with IPv4 and IPv6 connection whether I see using wireguard VPN IPv6 support (https://ipv6-test.com/) but it only shows "IPv6 not supported".

**Comment by MichaIng** [comment]
Hmm, indeed the Unique Local Unicast (fd00::) and Global Unicast (200x::) addresses seem to be doubled somehow. Now sure if/how this is related to the WireGuard PostUp commands as those should only have an effect on how incoming requests are handled and forwarded. Those addresses should be assigned by the router. How do you do that, DHCPv6?

On the Pi check whether the rules have been applied successfully and whether WireGuard reported any errors:
```
sysctl net.ipv6.conf.eth0.accept_ra net.ipv6.conf.{eth0,wg0}.forwarding
journalctl -u wg-quick@wg0
```

**Comment by ovz93br43v7** [comment]
> Hmm, indeed the Unique Local Unicast (fd00::) and Global Unicast (200x::) addresses seem to be > doubled somehow. Now sure if/how this is related to the WireGuard PostUp commands as those > > should only have an effect on how incoming requests are handled and forwarded. 

I have restarted the RPi and after that the double addresses are gone. Very curious :-/ 
EDIT: Some minutes later `ip a` shows the same like above (double addresses)
I don't know whether it helps to narrow this down but I removed for testing purpose the 4 PostUp/PostDown commands from the wireguard config. After reboot and since ~12 hours `ifconfig` / `ip a` shows just three ipv6 addresses (fd00, fe80, 200x).

> Those addresses should be assigned by the router. How do you do that, DHCPv6?

I use a Fritzbox as a router and yes DHCPv6 is activated but at the moment only IPv6 DNS Server is assigned to the clients. There are two other options (assign DNS+IA_PD or assign DNS+IA_PD+IA_NA) but according to some threads in the Pihole forum the most common advise is there to use only the first option "assign DNS-Server". 
Would you recommend one of the other options?

> On the Pi check whether the rules have been applied successfully and whether WireGuard reported any errors:
> 
> ```
> sysctl net.ipv6.conf.eth0.accept_ra net.ipv6.conf.{eth0,wg0}.forwarding

Here the result:
```
net.ipv6.conf.eth0.accept_ra = 2
net.ipv6.conf.eth0.forwarding = 1
net.ipv6.conf.wg0.forwarding = 1
```
> journalctl -u wg-quick@wg0

Here the result:
`````````````
-- Logs begin at Sun 2020-06-14 23:01:17 CEST, end at Sun 2020-06-14 23:02:14 CEST. --
Jun 14 23:01:43 raspberrypi systemd[1]: Starting WireGuard via wg-quick(8) for wg0...
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] ip link add wg0 type wireguard
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] wg setconf wg0 /dev/fd/63
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] ip -4 address add xxx.xxx.xxx.xxx/24 dev wg0
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] ip link set mtu 1420 up dev wg0
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] sysctl net.ipv6.conf.eth0.accept_ra=2
Jun 14 23:01:43 raspberrypi wg-quick[707]: net.ipv6.conf.eth0.accept_ra = 2
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] sysctl net.ipv6.conf.wg0.forwarding=1 net.ipv6.conf.eth0.forwarding=1
Jun 14 23:01:43 raspberrypi wg-quick[707]: net.ipv6.conf.wg0.forwarding = 1
Jun 14 23:01:43 raspberrypi wg-quick[707]: net.ipv6.conf.eth0.forwarding = 1
Jun 14 23:01:43 raspberrypi wg-quick[707]: [#] ip6tables -A FORWARD -i wg0 -j ACCEPT; ip6tables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
Jun 14 23:01:43 raspberrypi systemd[1]: Started WireGuard via wg-quick(8) for wg0.
`````````````

**Comment by MichaIng** [comment]
IA_PD+IA_NA should not be required indeed.

I was reading a bid further into it and get contrary info. In some cases an IPv6 subnet is used additionally for the tunnel network: https://stanislas.blog/2019/01/how-to-setup-vpn-server-wireguard-nat-ipv6/
This somehow makes sense as every adapter that wants to communicate via IPv6 requires an IPv6 address. However the last time I tested it I was able to ping IPv6 addresses without it and some guides also do not contain this while forcing IPv6 tunnelling (client-side) and forwarding (server-side): https://wiki.ubuntuusers.de/WireGuard/
In IPv6 many things  are automated and probably that additional two IPv6 addresses on the `eth0` interface are due to forwarding IPv6 from the IPv4-only wg0 interface. Adding an IPv6 subnet to the WireGard network may compensate this or in your case enable IPv6 through the VPN at all.

Probably some network specialist needs to give clarification, probably I find some time do run tests here and verify that/how it works in my case without assigning IPv6 addresses to the WireGuard interfaces 🤔.

**Comment by ovz93br43v7** [comment]
@MichaIng Many thanks for your effort! I'm not at all a network specialist but I can support with additional tests in my configuration if needed and if you want. When I look in your links and in this link from ubuntuusers: https://technofaq.org/posts/2017/10/how-to-setup-wireguard-vpn-on-your-debian-gnulinux-server-with-ipv6-support/ then I see there is a IPv6 address defined in the /etc/wireguard/wg0.conf interface section. In my config file is at the moment only a IPv4 address defined. But is this a random generated IPv6 address?!? :-/

**Comment by MichaIng** [comment]
The Ubuntu wiki does not use an IPv6 address but the other one does and yes this is the first thing I'd try as well.

> But is this a random generated IPv6 address?!?

Not totally random, but it seems to be an own chosen one from the ULA (unique local address) range.
I found another source where they used a sub range from the one that the router assigned to the eth0 interface, in your case one from the `fd00::XXX` address with e.g. `/72` mask (so a smaller range). However actually it should not matter as long as it does not collide with the other VPN host or clients IPv6 addresses, so choosing an own one should work fine.

**Comment by btb23** [comment]
Thanks to the previous comments and tips, with a little addition to the configs I have ipv6 fully working on the wg tunnel, also works with a cellular connection which is only ipv4.

Here are my working configs.

wg0.conf

```[Interface]
PrivateKey = hide
Address = 10.6.0.1/24,fd42:42:42::1/64
ListenPort = 51899
PostUp = sysctl net.ipv6.conf.eth0.accept_ra=2
PostUp = sysctl net.ipv6.conf.%i.forwarding=1 net.ipv6.conf.eth0.forwarding=1
PostUp = ip6tables -A FORWARD -i %i -j ACCEPT; ip6tables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
PostDown = ip6tables -D FORWARD -i %i -j ACCEPT; ip6tables -t nat -D POSTROUTING -o eth0 -j MASQUERADE

[Peer]
PublicKey = hide
PresharedKey = hide
AllowedIPs = 10.6.0.2/32, fd42:42:42::2/128
```

And the client config

```[Interface]
PrivateKey = client 1 private key
Address = 10.6.0.2/24,fd42:42:42::2/64
DNS = 10.6.0.1

[Peer]
PublicKey = server public key
Endpoint = <server public IP>:51899
AllowedIPs = 0.0.0.0/0,::/0
```
And the connection is up

wg show

```interface: wg0
  public key: hide
  private key: (hidden)
  listening port: 51899

peer: hide
  preshared key: (hidden)
  endpoint: xxx.xxx.xxx.xxx:40869
  allowed ips: 10.6.0.2/32, fd42:42:42::2/128
  latest handshake: 8 seconds ago
  transfer: 335.25 KiB received, 1.40 MiB sent
```
To add another peer simply increment the last digit in the ipv6 address as we do with ipv4

I do like the simplicity of using pivpn to add wireguard tunnels, with these few additions ipv6 is fully working too, well for me, anyone else find this a solution for them?

PS Edit. I thought I had the code blocks above, but failed there!

**Comment by mwoolweaver** [comment]
@btb23 use 3  ```

 ``` 
to signify

multi line 

code blocks
 ```


use 1 `

`to signify single line code`

**Comment by ovz93br43v7** [comment]
@btb23  thanks for sharing your solution. I have now found the time to try that on my pi. In my first trails I had no success.
Issues on my side after trying to transfer your solution into my configuration:
1. Not really a problem but confusing: `ip a` shows on `eth0` two IPv6 addresses.
2. `wg show` shows me also the configured IPv6 address but all IPv6 tests are not successful. (cellular connection with IPv4 and WiFi with IPv4/6)

Questions regarding your solution:
1. Is this address `fd42:42:42::1/64` just random or according to the ULA defined in your router? I have tried both router ULA +1,2,3,4 for the clients and with a random ULA.
2. Why one time `/64` and for the peers `/128`? (I tried both: `/64` for all, `/128` only for clients)
3. Have you activated port forwarding for IPv4 and IPv6 in your router?
4. Have also changed something at your clients? I assume that the configuration in the android / iOS app has to be updated or not?

**Comment by BakedCrossiant** [comment]
is there any plans to add listening on IPv6 as a non-default?

**Comment by Dunrar** [comment]
I'm a (potential) user of PIVPN with Wireguard from germany. So far, I had no luck getting it to work, though. 

My ISP uses DS-Lite, so I can't connect via IPv4 without jumping through some hoops. This seems to be standard practice now, at least in germany. Realistically, this leaves only IPv6 for many if not most. 

Luckily, my mobile network provider already provides IPv6. This is not as common yet, but probably will be in the near future. Even so, using the aforementioned changes to the config files and the system, I was not able to get any form of connection to the VPN. 

I'll probably follow a tutorial to install Wireguard by hand now, instead of trying to change the pre-configured config files. This will most likely be educational, but I love PIVPNs goal of making the setup and management of VPNs more feasible for many.

Seeing as I would consider mine a fairly standard scenario, I would like to cast my vote in favor of some form of official/optional IPv6 support. Thank you!

**Comment by orazioedoardo** [comment]
@Dunrar I would do that but I'm unable to test as I do not have access to (multiple) IPv6 networks, don't know about @4s3ti. If someone is willing to make changes and test, why not? We could add support.

**Comment by coolapso** [comment]
unfortunately I have no reliable way to test IPv6 =/

**Comment by MichaIng** [comment]
I would be able to run test. I could also open a related PR, but it could take a while until I find time, so would be great if someone else could do that,

A working config for WireGuard has been posted above: https://github.com/pivpn/pivpn/issues/259#issuecomment-699619034

**Comment by Anuskuss** [comment]
I can't get this to work, but I'm fortunate enough to have true DualStack, so I atleast have IPv4 going for me. I don't even really understand why I can't use IPv6 through my IPv4 tunnel though. Do I really need two separate connections, one to my public IPv4 and one to my public IPv6 address, just to have support for both types? That's sounds kinda wasteful IMO but I guess that's just how it works. Hopefully PiVPN will sooner than later make this whole process as painless as possible since the world is heading to IPv6.

**Comment by MichaIng** [comment]
> I don't even really understand why I can't use IPv6 through my IPv4 tunnel though. Do I really need two separate connections, one to my public IPv4 and one to my public IPv6 address, just to have support for both types?

The connection to the VPN server is not relevant for the connections passed through the VPN server. So you can without problems connect via IPv4 to the VPN server and then have an IPv6 connection to a remote host tunnelled through it.

If a client is in an environment with IPv6 support and has it not actively disabled, it likely has a GUA IPv6 address as well and will in most cases by default try to connect via IPv6 to hosts if an AAAA record is provided. In this case, you don't want to bypass your VPN ("IPv6 leak"), but want to have that IPv6 request tunnelled as well, and successful instead of hanging. The client VPN config needs to force all IPv6 requests through the tunnel (which is already the case with the default PiVPN client config, if I'm not mistaken) and the VPN server needs to support IPv6, have IPv6 forwarding and NATing enabled, which is covered by the `sysctl` and `iptables6` commands provided above.

**Comment by Anuskuss** [comment]
@MichaIng Okay good, I didn't really invest too much time because I couldn't even establish a connection to my public IPv6 address (port was open and firewall deactivated for the host) but if this can already work. I'll give it a try tomorrow.
Although I'd like to know this one as well:
> Is this address `fd42:42:42::1/64` just random or according to the ULA defined in your router? I have tried both router ULA +1,2,3,4 for the clients and with a random ULA.

**Comment by MichaIng** [comment]
That is the VPN internal IPv6 address, so any address (range) within the `fd00::/8` block will work. Of course clients then need to be defined within the same subnet.

I read again through the old posts above, and indeed there seem to be cases where the dedicated/additional IPv6 addresses on the WireGuard network seem to be required. At least worth to test when it's not working without, adding that to server and client configs shouldn't be so hard. I'll also play around with it when I find time.

**Comment by Anuskuss** [comment]
> That is the VPN internal IPv6 address, so any address (range) within the `fd00::/8` block will work.

I know that these are ULAs, but I didn't know that any device can just claim one (I though that only the DHCPv6 can). Anyway, I copied [**btb23**'s config](https://github.com/pivpn/pivpn/issues/259#issuecomment-699619034) exactly, but that didn't work for me (I couldn't even ping `fd42:42:42::1`). It's probably the iptables that are not configured correctly in my setup, but I honestly don't care. I chose PiVPN because I didn't want to deal with this, so let's hope the devs provide a automagical solution 😛

**Comment by MichaIng** [comment]
> I didn't know that any device can just claim one (I though that only the DHCPv6 can).

Well, every device can apply itself any IP (or you can apply any IP to any device). If a DHCPv6 server is present, than the device can ask which ULA it shall apply and respect it, or keep looking for RAs to do SLAAC do something totally different. Not that it would make sense, but it's the devices which apply the IPs themselves and nothing external can strictly force them to do anything 😉.

In case of the VPN, there is no DHCP-ish server, but the client indeed must create the network interface, apply the IP and the routes everything themselves, which is practically done by the `wg-quick` command/service, respectively the WireGuard client, based on the config you gave them. Then it can try to establish the connection with the server and if IP and key matches to what the server has stored about that client, it is granted.

I cope you kept your private and public key info when applying btb23's configs 😉. To debug, check the WireGuard server logs and whether IP and routes have been applied as expected:
```sh
ip a s wg0
ip r l dev wg0
ip -6 r l dev wg0
```

**Comment by Anuskuss** [comment]
@MichaIng Appreciate the support, but I don't think just helping me to pull this off is gonna amount to much. Instead you should make it work on your setup first, create a PR and hope to get enough input from the PiVPN community (including me) to get something going that works for everyone :)

**Comment by MichaIng** [comment]
Well it might help to answer the last open questions before creating the PR. But yes, often having a PR is a more productive basis for testing and discussion. I'll see if I find time, as currently I'm very busy with own projects.

**Comment by DerDanilo** [comment]
For the project "wg-easy" IPv6 is supported (https://github.com/WeeJeWel/wg-easy/pull/191) and we should be able also bring this to pivpn.
There is a lot of information on IPv6 for pivpn. So what is the current state here? I am happy to contribute but would like to know if the time I might invest will be worth it. If done right, will the PR for IPv6 be merged?

**Reply by DerDanilo** [reply]
Here we go: https://github.com/pivpn/pivpn/pull/1455

**Reply by coolapso** [reply]
@DerDanilo I don't see any reason to not merge it ... the main blocker with this so far is because no one really has a way to test and implement it, I am more than glad to review the PR and merge it provided there is no comments, if there is I am sure we can work them out! :) 

I Might be able to help testing it, but i am not sure yet.
